query_optimizer/OptimizerTree.hpp

```diff
@@ -118,6 +122,169 @@ class OptimizerTree : public OptimizerTreeBase {
+        : key_(key), value_(value) {  }
+
+    std::string getName() const {
+      return "ProtoProperty";
+    }
+
+    void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                             std::vector<std::string> *inline_field_values,
+                             std::vector<std::string> *non_container_child_field_names,
+                             std::vector<TreeNodeType> *non_container_child_fields,
+                             std::vector<std::string> *container_child_field_names,
+                             std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+      inline_field_names->push_back(std::string("Property"));
+      inline_field_values->push_back(key_);
+
+      inline_field_names->push_back(std::string("Value"));
+      inline_field_values->push_back(value_);
+    }
+
+   private:
+    std::string key_;
+    std::string value_;
+  };  // class OptimizerProtoRepresentation
+
+ public:
+  OptimizerProtoRepresentation() : properties_() { }
+
+  std::string getName() const {
+    return "ProtoDescription";
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value A string value.
+   */
+  void addProperty(std::string key, std::string value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, value));
+    properties_.push_back(property);
+  }
+
+  /**
+   * @brief Add a key-value type property to this proto representation.
+   * @details Internally, the instance creates another node to represent the
+   *          given values.
+   *
+   * @param key A string key.
+   * @param value An int value which is converted into a string.
+   */
+  void addProperty(std::string key, int value) {
+    std::shared_ptr<const OptimizerProtoPropertyRepresentation> property(
+        new OptimizerProtoPropertyRepresentation(key, std::to_string(value)));
+    properties_.push_back(property);
+  }
+
+  void getFieldStringItems(std::vector<std::string> *inline_field_names,
+                           std::vector<std::string> *inline_field_values,
+                           std::vector<std::string> *non_container_child_field_names,
+                           std::vector<TreeNodeType> *non_container_child_fields,
+                           std::vector<std::string> *container_child_field_names,
+                           std::vector<std::vector<TreeNodeType>> *container_child_fields) const {
+    for (auto property : properties_) {
+      non_container_child_field_names->push_back("Property");
+      non_container_child_fields->push_back(property);
+    }
+  }
+  // A list of managed properties.
+  std::vector<std::shared_ptr<const OptimizerProtoPropertyRepresentation> > properties_;
```

Comment:
Ah, I see. The vector calls the destructors on all of its stored objects, which are unique_ptr in this case. And the unique pointer calls the destructors to the pointed objects.